### PR TITLE
1st Draco_Mesh ghost cell implementation

### PR DESCRIPTION
**Pull request by RyanWollaeger**
### Background

* Domain decomposed meshes will generally need to supply some ghost cell information to solvers.
* This update assumes the ghost cell data can be supplied in a particular form to the mesh ctor.

### Purpose of Pull Request

* Add ghost cell data to `Draco_Mesh` ctor, to generate a ghost cell layout.
* Refactor some emergent, repeated functionality.
* [Towards Redmine Issue #1182](https://rtt.lanl.gov/redmine/issue/1182)

### Description of changes

* Add ghost cell layout to `Draco_Mesh`, generated by mesh ctor.
* Add 2-cell domain-decomposed unit test to test ghost cell layout generation.
* Refactor portion of repeated code for map generation in `compute_cell_to_cell_linkage` to one separate function.
* Refactor test mesh data generation into re-usable class for unit tests, `Test_Mesh_Interface`.
* Move layout serialization functions to members of `Test_Mesh_Interface`.

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [ ] Travis CI checks pass
  * [ ] Code coverage does not decrease
  * [ ] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [ ] Code reviewed/approved, sufficient DbC checks, testing, documentation

### CODE REVIEW

File: src/mesh/Draco_Mesh.hh [hunk 1]
```diff
@@ -113,12 +132,19 @@ private:
...
       const std::vector<unsigned> &cell_type,
       const std::vector<unsigned> &cell_to_node_linkage,
       const std::vector<unsigned> &side_node_count,
-      const std::vector<unsigned> &side_to_node_linkage);
+      const std::vector<unsigned> &side_to_node_linkage,
+      const std::vector<unsigned> &ghost_cell_type,
+      const std::vector<unsigned> &ghost_cell_to_node_linkage);
+
+  //! Calculate a map of node to vectors of indices (cells, sides, ghost cells)
+  std::map<unsigned, std::vector<unsigned>> compute_node_indx_map(
```
Comment (KineticTheory) [inline src/mesh/Draco_Mesh.hh]: Please prefix with DLL_PUBLIC_mesh

File: src/mesh/Draco_Mesh.cc [hunk 2]
```diff
@@ -256,21 +243,81 @@ void Draco_Mesh::compute_cell_to_cell_linkage(
...
+      const std::vector<unsigned> &vert0_ghosts =
+          node_to_ghost_cell_map[vec_node_vec[l][0]];
+      const std::vector<unsigned> &vert1_ghosts =
+          node_to_ghost_cell_map[vec_node_vec[l][1]];
+
+      // find common ghost cells
+      std::vector<unsigned> ghost_cells_in_common;
+      std::set_intersection(vert0_ghosts.begin(), vert0_ghosts.end(),
+                            vert1_ghosts.begin(), vert1_ghosts.end(),
+                            std::back_inserter(ghost_cells_in_common));
```
Comment (KineticTheory) [inline src/mesh/Draco_Mesh.cc]: I didn't know this function existed! Cool.

File: src/mesh/test/Test_Mesh_Interface.hh [hunk 3]
```diff
@@ -0,0 +1,238 @@
+//----------------------------------*-C++-*----------------------------------//
+/*!
+ * \file   mesh/test/Test_Mesh_Interface.cc
+ * \author Ryan Wollaeger <wollaeger@lanl.gov>
```
Comment (KineticTheory) [inline src/mesh/test/Test_Mesh_Interface.hh]: If this file gets any larger, we probably should generate a `libmesh_test.so` file and teach the unit tests to link to the test library. Don't do this yet - but think about it for future work.

File: src/mesh/Draco_Mesh.hh (hunk 1 above)
Comment (RyanWollaeger) [reply src/mesh/Draco_Mesh.hh]: This function is currently private.

File: src/mesh/test/Test_Mesh_Interface.hh (hunk 3 above)
Comment (RyanWollaeger) [reply src/mesh/test/Test_Mesh_Interface.hh]: This is a good point - I accidentally labeled the file as .cc in the header, possibly subconsciously preparing the test library!

File: src/mesh/Draco_Mesh.cc (hunk 2 above)
Comment (RyanWollaeger) [reply src/mesh/Draco_Mesh.cc]: It should be noted that this complexity is achieved as a result of the vector arguments being sorted, by construction.

**Pull request merged**